### ISSUE TITLE
RTVS Tooling for producing Dockerized Shiny Apps

### ISSUE THREAD

**Issue by whatevergeek** [Microsoft/RTVS#3535]
Since RTVS is already leveraging docker in the workspaces,
perhaps, can add a feature/tooling to produce dockerize shiny apps.

It'll be like what we have in aspnet core on docker.
This will help us deploy our shiny apps (running on docker) anywhere.
Feel free to comment if further detail is needed.

**Comment by Hong-Revo** [comment]
Wait, what? RTVS uses Docker?

**Comment by whatevergeek** [comment]
yes! i just saw John Lam's new presentation... 
so so awesome! he connected to a remote rtvs workspace on docker (rtvs-linux)
 this is my favorite feature to date...

**Comment by Hong-Revo** [comment]
Ah, I see. RTVS itself doesn't use Docker as far as i know, but you can certainly create a container that includes it.

**Comment by MikhailArkhipov** [comment]
We are still working on it. @jflam 's demo was very early result.

**Comment by MikhailArkhipov** [comment]
However, be aware that Shiny server is a proprietary product and we can't just produce containers with it. However, the container can encapsulate pre-set client environment which can be connected to and then run Shiny locally.

**Comment by whatevergeek** [comment]
it's ok.. i was able to pull the image from dockerhub and replicate his demo... so so cool... i love it...  when you say "still working on it", does it include tooling for making dockerized shiny apps? i'm think towards using rtvs for creating/developing data products :-)

btw, do you have the dockerfile for rtvs-linux?
i want to add my own set of R libraries :-)

**Comment by MikhailArkhipov** [comment]
Well, "_it'll be like what we have in aspnet core on docker. This will help us deploy our shiny apps (running on docker) anywhere_". - I'd appreciate some insights what exactly is the expectation :-). If something like a Shiny server, it container running on Azure you can connect to from a browser remotely, that is unlikely since it would entail re-implementing Shiny server on ASP.NET or something like that. Packaging existing server is not possible since it is proprietary product. However, presetting environment and then connecting to a container from another instance of RTVS as remote session certainly possible.

**Comment by whatevergeek** [comment]
if shiny is proprietary, are we planning to have something equivalent?
maybe marry a subset of powerbi?
so that we can have a self contained data product within a container that we can deploy anywhere...

**Comment by jflam** [comment]
We don't yet have a Dockerfile for our Linux container. @karthiknadig built it by hand, so all we have is an image. But we absolutely plan to:

1. Have Dockerfile + build for all of our images
1. Have full tooling support for container lifecycle management as well. What would you like to see here?

**Comment by whatevergeek** [comment]
yes to both
thanks John! :-)

**Comment by MikhailArkhipov** [comment]
@karthiknadig is looking into options here

**Comment by whatevergeek** [comment]
Thanks @MikhailArkhipov and @karthiknadig

Btw, i tried using Karthik's docker image previously as base... in my own custom dockerfile... seems to work... hence, i'm not making noises after that... hehehe... :-) 

On the otherhand, i saw Sirosh's talk on AI few days ago. He talked about deploying models as containers. This is similar to what i have in mind when i posted this previously.

Anyway,  feel free to let me know if there's anything I can help and thanks again for looking into this. :-)

**Comment by smortaz** [comment]
i agree with paul - the holy grail - which joseph talked about, has been on
our radar forever.  we just havent had the backend machinery to build a UI
against yet.  with AML v2.0 some of these will come to fruition and
rtvs/ptvs can build proper tooling for them.

**Comment by karthiknadig** [comment]
From prototype experimentation, I was able to docker build a image that would run shiny application. So it should be possible to automate this from RTVS client. However, there is more work needed to figure out the end-to-end workflow, for the build and publish scenarios.

Note: Some R packages cannot be installed without the required OS package installed. For example, [curl](https://cran.r-project.org/web/packages/curl/index.html) package requires libcurl4-openssl-dev. So we will need to add these to the docker file as well.

**Comment by whatevergeek** [comment]
@karthiknadig end to end workflow would be great!  Thanks for the update! :-)

**Comment by Hong-Revo** [comment]
You may already be aware of this, but there are prebuilt R Dockers available at https://hub.docker.com/u/rocker/

Maybe you could leverage those?

**Comment by whatevergeek** [comment]
@Hong-Revo you mean using them as base images by @karthiknadig , right? 
'cause those images alone can't be used directly with Visual Studio Remote workspaces feature.
i'm using kvnadig/rtvs-linux for that...
btw, @karthiknadig i saw this on dockerhub: microsoft/rtvs

is this the new image?

**Comment by karthiknadig** [comment]
microsoft/rtvs images are the new set of images. These work with Container UI feature that was added in RTVS 1.3 in VS 2017 15.5 preview 2. Here are the docs to use that feature:
https://github.com/Microsoft/RTVS/blob/master/doc/rtvsd/rtvs-docker-container.md

Note: I will add instructions on how to create your own image and container and have it show up in RTVS UI.

**Comment by whatevergeek** [comment]
Wow! This is fresh.. VS2017 15.5 prev2 was just released yesterday.. didn't know this feature is there... will check it out... thanks @karthiknadig

**Comment by whatevergeek** [comment]
@karthiknadig  was keen to try... and so i downloaded the new preview version...
but felt so sad when i saw this:
![image](https://user-images.githubusercontent.com/7351683/31997724-93ac1aac-b9bf-11e7-8d07-e8f41fb43cdc.png)

it needs "docker for windows".
i can only run docker toolbox on my machine.
can we make this work also for docker toolbox?

**Comment by karthiknadig** [comment]
@whatevergeek If you can use the command line to manage the container, you can connect to it from RTVS. Although the RTVS Container UI will be disabled for you (that requires Docker for Windows). I recommend using this Dockerfile template. Note: this will only work with RTVS version 1.3
```
FROM microsoft/rtvs:1.3-ub1604-rlatest
RUN useradd --create-home ruser1
RUN echo "ruser1:foobar" | chpasswd
```
The above container will have R 3.4.2, rtvs-daemon, and a bunch of R packages pre-installed.

Build and run the container:
```
docker build -t my-rtvs-image:latest .
docker run -p 6056:5444 my-rtvs-image:latest rtvsd
```

Enter connection info like this to add remote connection. Note the port number used in the previous step and here.
![image](https://user-images.githubusercontent.com/3840081/32014216-4f3bec82-b972-11e7-81d9-03936a082a7a.png)


Enter username and password like this. Remember to add the `<<unix>>\` prefix.
![image](https://user-images.githubusercontent.com/3840081/32014123-0e999fda-b972-11e7-9650-8cd0f0a02d27.png)

Sadly, I don't think we can support docker toolbox based on the development cost. Also, since docker recommends moving to docker for windows. These are some other options you could try:
https://github.com/Microsoft/RTVS/blob/master/doc/rtvsd/rtvs-daemon-installation.md
Since you are working with VS2017 preview 2, change the rtvs-daemon to this version:
```
wget -O rtvs-daemon.tar.gz https://github.com/Microsoft/RTVS-Docker/blob/master/1.3/ub1604/base/rtvs-daemon_1.3.31011.1002.tar.gz?raw=true
```

**Comment by whatevergeek** [comment]
Thanks @karthiknadig
I'll try as advised

**Comment by whatevergeek** [comment]
hi @karthiknadig 
i'm unable to access https://github.com/Microsoft/RTVS-Docker
is it a private repo?

![image](https://user-images.githubusercontent.com/7351683/32029224-bd689968-ba25-11e7-8a4c-c910bafafda2.png)

**Comment by karthiknadig** [comment]
@whatevergeek Forgot about the private repo. here is the path from my repo:
https://github.com/karthiknadig/docker-stuff/blob/master/rtvs-daemon_1.3.31011.1002.tar.gz?raw=true

**Comment by whatevergeek** [comment]
Thanks @karthiknadig
I'll check accordingly

**Comment by whatevergeek** [comment]
Thanks @karthiknadig
I'll check accordingly

**Comment by whatevergeek** [comment]
hi @karthiknadig 
i'm trying your tutorial on using docker container for RTVS:
https://github.com/Microsoft/RTVS/blob/master/doc/rtvsd/rtvs-docker-container.md

After clicking "Create"... it's stuck here:
build> Successfully built 3509fe93496b
build> Successfully tagged 37a499bb-b42d-c43c-cf54-43da64048bf0:latest
build> SECURITY WARNING: You are building a Docker image from Windows against a non-Windows Docker host. All files and directories added to build context will have '-rwxr-xr-x' permissions. It is recommended to double check and reset permissions for sensitive files and directories.


Any other special setup I need to do?
I'm expecting the build container in the VS Container window:
![image](https://user-images.githubusercontent.com/7351683/32052062-96e4195c-ba89-11e7-81bb-bd0346c633d3.png)

**Comment by karthiknadig** [comment]
Can you do a `docker inspect`? on that container. In the resulting JSON, look for `PortBindings` under `HostConfig`. Can you paste the result? It should have at least one binding for port '5444/tcp'.

**Comment by whatevergeek** [comment]
hi @karthiknadig 
no container was created...
an image was created though... so i did a docker inspect on it:
it has 5444/tcp in the "ContainerConfig" and "Config"
            "ExposedPorts": {
                "5444/tcp": {}
            },

apart from "Create"
do i need to create something via "From Template" 
in the Containers Window?

**Comment by karthiknadig** [comment]
When you create a container create it like this:
```
docker create -p 6060:5444 --name myrtvscontainer 37a499bb-b42d-c43c-cf54-43da64048bf0:latest rtvsd
docker start  myrtvscontainer
```
I need to investigate why container creation does not happen after build. But lets make sure it can pickup a container after it is created.
